fl4health/utils/dataset.py

```diff
@@ -103,3 +102,55 @@ def __getitem__(self, index: int) -> Tuple[Dict[str, torch.Tensor], torch.Tensor
     def __len__(self) -> int:
         first_key = list(self.data.keys())[0]
         return len(self.data[first_key])
+
+
+class SyntheticDataset(TensorDataset):
+    def __init__(
+        self,
+        data: torch.Tensor,
+        targets: torch.Tensor,
+    ):
+        """
+        A dataset for synthetically created data strictly in the form of pytorch tensors.
+        Args:
+            data (torch.Tensor): Data tensor with first dimension corresponding to the number of datapoints
+            targets (torch.Tensor): Target tensor with first dimension corresponding to the number of datapoints
+        """
+        assert data.shape[0] == targets.shape[0]
+        self.data = data
+        self.targets = targets
+
+    def __getitem__(self, index: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        assert self.targets is not None
+
+        data, target = self.data[index], self.targets[index]
+        return data, target
+
+    def __len__(self) -> int:
+        return len(self.data)
+
+
+D = TypeVar("D", bound=Union[TensorDataset, DictionaryDataset])
+
+
+def select_by_indices(dataset: D, selected_indices: torch.Tensor) -> D:
+    if isinstance(dataset, TensorDataset):
+        modified_dataset = copy.deepcopy(dataset)
```

Comment:
I think the reason to do deepcopy here because it would get a new memory than directly changing input dataset, so we can apply selection multiple times and wouldn't modify the original one. But should not we do the same for DictionaryDataset as well in line 146?

Reply:
I'm not 100% sure. So correct me if I'm wrong here. However, in the original implementation of `select_by_indices` we were making modifications to the `dataset.targets` and `dataset.data` in-place, which resulted in changes to the original object. Here, we're creating a new DictionaryDataset with these values on line 154. So the original dataset should still be intact and unchanged.

Reply:
Oh I see in the last line we do that. Yes yes then it should be fine.